### ISSUE TITLE
Wrong typings generation

### ISSUE THREAD

**Issue by RIP21** [formium/tsdx#967]
### Current Behavior
When generates d.ts from this source
```typescript
import { useToast as useChakraToast } from '@chakra-ui/react'
import { useMemo } from 'react'
import type { UseToastOptions } from '@chakra-ui/react'

export const useToast = (options?: UseToastOptions) => {
  const defaultOptions = useMemo(() => {
    return {
      position: 'top-right' as UseToastOptions['position'],
      status: 'success' as UseToastOptions['status'],
      isClosable: true,
      ...(options ?? {}),
    }
  }, [options])
  return useChakraToast(defaultOptions)
}
```
Outputs this:
```typescript
export declare const useToast: (options?: any) => any;
```
### Expected behavior
Should output (and this is what regular `tsc` with `emitDeclarationsOnly` does):
```typescript
import type { UseToastOptions } from '@chakra-ui/react';
export declare const useToast: (options?: UseToastOptions | undefined) => {
    (options?: UseToastOptions | undefined): string | number | undefined;
    close: (id: string | number) => void;
    closeAll: (options?: import("@chakra-ui/react").CloseAllToastsOptions | undefined) => void;
    update(id: string | number, options: Pick<UseToastOptions, "status" | "render" | "position" | "duration" | "onCloseComplete" | "title" | "description" | "isClosable" | "variant">): void;
    isActive: (id: string | number) => boolean | undefined;
};
```

### Additional context

It's Rush.js managed monorepo so it uses hard links to link modules into `node_modules`.
I'm using TS 4. A similar workaround mentioned here https://github.com/formium/tsdx/issues/926 (using PNPM fashioned resolutions) is applied .

### Your environment

```text
  System:
    OS: Linux 5.3 Ubuntu 20.04.1 LTS (Focal Fossa)
    CPU: (16) x64 AMD Ryzen 7 3700X 8-Core Processor
    Memory: 449.17 MB / 31.37 GB
    Container: Yes
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 14.15.4 - ~/.nvm/versions/node/v14.15.4/bin/node
    Yarn: 1.22.10 - ~/.nvm/versions/node/v14.15.4/bin/yarn
    npm: 6.14.10 - ~/.nvm/versions/node/v14.15.4/bin/npm
  Browsers:
    Chrome: 87.0.4280.88
    Firefox: 85.0

```

**Comment by RIP21** [comment]
Unfortunately, there is no way to disable `d.ts` generation completely to replace this functionality with regular `tsc` into the same `dist` folder as `tsdx` does.

**Comment by RIP21** [comment]
The only workaround I found for now is
```
    "start": "tsdx watch --transpileOnly --onSuccess \"tsc --emitDeclarationOnly --noEmit false --outDir dist/types\"",
    "build": "tsdx build --transpileOnly && tsc --emitDeclarationOnly --noEmit false --outDir dist/types",
```
And typings should point into a different spot too.
```
 "typings": "dist/types/index.d.ts",
```
Although takes a bit more time.

**Comment by jessekrubin** [comment]
@RIP21 What version of rush and pnpm are you using?

**Comment by RIP21** [comment]
> @RIP21 What version of rush and pnpm are you using?
  "rushVersion": "5.38.0"
  "pnpmVersion": "5.15.2"

**Comment by jessekrubin** [comment]
@RIP21 I am using the same versions. I often get very strange type errors when building react component libs. Do you?

**Comment by RIP21** [comment]
> @RIP21 I am using the same versions. I often get very strange type errors when building react component libs. Do you? 

Not really. If you, by accident, have, preserveSymlinks: true in tsconfig, make sure to disable it that will solve most of the problems.
But other than that, should be fine.

**Comment by jessekrubin** [comment]
@RIP21 Hm. Just checked and I do not have that set. All the errors I get are of the form 'propertyName' does not exist on type 'IntrisicAttributes & ... etc`

Here is an error message I get when using a `Stack` from `@fluentui/react`: `Property 'tokens' does not exist on type 'IntrinsicAttributes & IStackItemProps & { children?: ReactNode; }'`

I get a similar error using material-ui as well as several other react-component libs.

**Comment by jessekrubin** [comment]
@RIP21 I don't get those types of errors when building outside of rush/pnpm. I can spread in the props and it sometimes goes away but it's inconsistent

**Comment by agilgur5** [comment]
So #953 reports part of the same error (the "no member" error) similarly in a Rush monorepo. It seemed like it might be due to the PNPM hard links that things break, but I haven't much of a clue why. I did a bit of debugging in that issue but really not sure where to go after those simpler leads turned up nothing. It's not clear why hard links would break things.

TSDX uses `rollup-plugin-typescript2` for practically all of the TS processing, so my guess is the error is upstream there. If you're able to reproduce solely with that one plugin (I had been working on a CodeSandbox for repros for rpts2 given how frequently they happen and the amount of config TSDX wraps that can make a minimal repro difficult, but never quite got to it -- maybe I'll do that soon), that would narrow it down to rpts2 as the culprit and can file an upstream issue there.
There's a chance that some other of the Rollup plugins we use is responsible though, e.g. `resolve` or something struggling with hard links, so failure to reproduce with plain rpts2 would point in that direction

**Comment by agilgur5** [comment]
For reference, I am fairly sure I fixed this upstream in https://github.com/ezolenko/rollup-plugin-typescript2/pull/332, which fixed https://github.com/ezolenko/rollup-plugin-typescript2/issues/330, which is a very similar downstream issue with `pnpm` symlinks.

To use that in your project, you can set your [`pnpm` `overrides`](https://pnpm.io/package_json#pnpmoverrides) to use `rollup-plugin-typescript2` [0.32.0+](https://github.com/ezolenko/rollup-plugin-typescript2/releases/0.32.0):

```json
{
  "pnpm": {
    "overrides": {
      "rollup-plugin-typescript2": "^0.32.0"
    }
  }
}
```

And indeed, while the issue is _different_ here, the _root cause_ is the same as #953 so they are effectively duplicates.

> @agilgur5 what would you swap it out for? Esbuild? Some other thing?

No, I meant a different Rollup TypeScript plugin. They all have their own issues however and I was already contributing to rpt2 a bit in the past to fix upstream issues from TSDX. I found rpt2 to be the simplest to contribute to by a wide margin, so I actually recently stepped up to help maintain it and have fixed a ton of issues in the past month, including the most long-standing ones like this one with `pnpm` symlinks. Can see https://github.com/ezolenko/rollup-plugin-typescript2/issues/234#issuecomment-1139202740 for more details on that (and a root cause analysis on this and why TypeScript integrations are difficult to implement due to TS having a sparsely documented API).

**Comment by RIP21** [comment]
@agilgur5 thank you for your work! I'm no longer working on this codebase to try, but still. Tsdx is my tool to go, so I'm happy it's probably fixed now.
